src/refiner/refiner_util.fun

```diff
@@ -75,18 +75,35 @@ struct
            theta $ _ => GeneralRules.Unfolds (world, [(theta, NONE)]) goal
          | _ => raise Refine
 
-    fun CutLemma (world, theta : operator) =
+    fun CutLemma (world, theta : operator, oname) =
       let
         val {statement,...} = Development.lookupTheorem world theta
         val H >> P = statement
         val _ = if Context.eq (H, Context.empty) then () else raise Fail "nonempty context"
-        val name = Syntax.Variable.named (Syntax.Operator.toString theta)
+        val name = case oname of SOME name => name | NONE => Syntax.Variable.named (Syntax.Operator.toString theta)
       in
         GeneralRules.Assert (P, SOME name)
           THENL [GeneralRules.Lemma (world, theta), ID]
       end
-  end
 
+    structure CttCalculusView = RestrictAbtView
+      (structure Abt = Syntax and Injection = CttCalculusInj)
+
+    fun WfLemma (world, theta : operator) =
+      let
+        val name = Syntax.Variable.named (Syntax.Operator.toString theta)
+        val hyp = HypSyn.NAME name
+        val cutWfLemma = CutLemma (world, theta, SOME name)
+        val useHyp = GeneralRules.Hypothesis hyp ORELSE BHypRules.BHyp hyp
+        open CttCalculusInj CttCalculus CttCalculusView
+      in
+        fn goal as (_ |: H >> P) =>
+          (case project P of
+               MEM $ _ => cutWfLemma THEN useHyp
+             | EQ $ _ => cutWfLemma THEN GeneralRules.Unfolds (world, [(`> MEM, NONE)]) THEN useHyp
```

Comment:
The trick is, the lemma that we're cutting in has `MEM` in it, when the goal is `EQ`, so we have to unfold it before it will unify.

That reminds me—we should add the ability to mark `unfold` and `reduce` with targets (either the conclusion, or a hypothesis).